src/manager/TickManager.ts

```diff
@@ -1,46 +1,190 @@
-import JSBI from 'jsbi';
+import JSBI from "jsbi";
+import assert from "assert";
 import { Tick } from "../model/Tick";
 import { jsonMapMember, jsonObject } from "typedjson";
 
 @jsonObject
 export class TickManager {
   @jsonMapMember(Number, Tick, { name: "ticks_json" })
+  // key = Tick.tickIndex
   private ticks: Map<number, Tick>;
 
   constructor(ticks: Map<number, Tick> = new Map()) {
     this.ticks = ticks;
   }
 
   get(tickIndex: number): Tick {
-    return this.ticks.get(tickIndex) || new Tick();
+    if (this.ticks.has(tickIndex)) return this.ticks.get(tickIndex)!;
+    const newTick = new Tick(tickIndex);
+    this.set(newTick);
+    return newTick;
   }
 
-  set(tickIndex: number, tick: Tick) {
-    this.ticks.set(tickIndex, tick);
+  set(tick: Tick) {
+    this.ticks.set(tick.tickIndex, tick);
   }
 
-  flipTick(tickIndex: number, tickSpacing: number) {
-    // TODO
+  private nextInitializedTick(
+    sortedTicks: readonly Tick[],
+    tick: number,
+    leftToRight: boolean
+  ): Tick {
+    if (leftToRight) {
+      assert(!this.isBelowSmallest(sortedTicks, tick), "BELOW_SMALLEST");
+      if (this.isAtOrAboveLargest(sortedTicks, tick)) {
+        return sortedTicks[sortedTicks.length - 1];
+      }
+      const index = this.binarySearch(sortedTicks, tick);
+      return sortedTicks[index];
+    } else {
+      assert(
+        !this.isAtOrAboveLargest(sortedTicks, tick),
+        "AT_OR_ABOVE_LARGEST"
+      );
+      if (this.isBelowSmallest(sortedTicks, tick)) {
+        return sortedTicks[0];
+      }
+      const index = this.binarySearch(sortedTicks, tick);
+      return sortedTicks[index + 1];
+    }
   }
 
   getNextInitializedTick(
     tick: number,
     tickSpacing: number,
     leftToRight: boolean
   ): { nextTick: number; initialized: boolean } {
-    // TODO
-    return { nextTick: 0, initialized: false };
+    const sortedTicks = this.getSortedTicks();
```

Comment:
This step costs a lot of time for capacity of ticks, and swap as caller of this method, is considered of multiple rounds, so I think we can cache the sorted ticks since initializing state of ticks won't get updated during a swap process. This may mean we move `sortedTicks` to the input of `getNextInitializedTick`. However, it also bring complexity to API of `TickManager` as well as do no good to encapsulation. As to user case, swap is more frequent than management of liquidity position. It's better to use and maintain sortedArray than a map in `TickManager` from my perspective.

Reply:
I have thought about this as well, the issue with caching is that it would become expensive for initializing new ticks. Also, I would imagine the number of the elements to sort actually has an upper limit instead of growing linearly indefinitely. And since this is an optimization problem, how about keeping it as is, and wait for the bottleneck to show up first?

Reply:
Per our discussion, let's keep the map sorted upon insert and deletion instead of searching and update.

Reply:
Agreed